gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ProcessWorkUnitsWorkflowImpl.java

```diff
@@ -72,10 +73,33 @@ private CommitStats performWork(WUProcessingSpec workSpec) {
     searchAttributes = TemporalWorkFlowUtils.generateGaasSearchAttributes(jobState.getProperties());
 
     NestingExecWorkflow<WorkUnitClaimCheck> processingWorkflow = createProcessingWorkflow(workSpec, searchAttributes);
-    int workunitsProcessed =
-        processingWorkflow.performWorkload(WorkflowAddr.ROOT, workload, 0, workSpec.getTuning().getMaxBranchesPerTree(),
-            workSpec.getTuning().getMaxSubTreesPerTree(), Optional.empty());
-    if (workunitsProcessed > 0) {
+
+    Optional<Integer> workunitsProcessed = Optional.empty();
+    try {
+     workunitsProcessed = Optional.of(processingWorkflow.performWorkload(WorkflowAddr.ROOT, workload, 0,
+          workSpec.getTuning().getMaxBranchesPerTree(), workSpec.getTuning().getMaxSubTreesPerTree(), Optional.empty()));
+    } catch (Exception e) {
+      log.error("Exception occurred in performing workload,proceeding with commit step", e);
+      // We want to mark the GaaS flow as failure, in case performWorkFlow fails, but we still want to go ahead with commiting the workunits which were processed before failure
+      sendFailureEventToGaaS(workSpec);
+      return proceedWithCommitStepAndReturnCommitStats(workSpec, searchAttributes, workunitsProcessed);
```

Comment:
I was about to suggest you omit this line and "fall through", but the timing is actually crucial.  what you're doing here results in sending a failure GTE before doing the commit.  that would lead GaaS in turn to emit the job observability event *and* complete the flow execution before commit even runs.  if that case, the datasetsMetrics definitely won't be filled in the job observability event.

much better would be to wait to send the JobFailed GTE only *after* the JobSummary GTE is [sent](https://github.com/apache/gobblin/blob/4df6a52d520b4c8718a5df5c71dcef1248b1161e/gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/CommitStepWorkflowImpl.java#L63)

hence:
```
performCommitIfAnyWorkUnitsProcessed(workSpec, searchAttributes, workunitsProcessed);
throw e;  // [see comment for L90 about why we throw rather than sending a GTE directly]
```